Lightning Web Components

erocheleau opened the pull request:
SFINT team's experience with LWC.

malaporte commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+---
+
+In 2018, Salesforce introduced a new programming model for building Lightning components (Lightning components are the building blocks used to create applications within Salesforce). This new framework was designed with respect to the new web standards in order to allow web developers to keep using the knowledge they have acquired over the years outside of Salesforce. This new building model is a ...
```
```suggestion
In 2018, Salesforce introduced a new programming model for building Lightning components. As you might know already, Lightning components are the building blocks used to create applications within Salesforce. This new framework was designed with respect to the new web standards in order to allow web developers to keep using the knowledge they have acquired over the years outside of Salesforce. This new building model is aligned with the Web Components standard from the World Wide Web Consortium (W3C) and is using native browser features and language such as pure HTML and JavaScript. This blog post will document our team's experience with this new framework to build the Coveo for Salesforce integration using Lightning Web Components (LWC).
```

malaporte commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+<!-- more -->
+
+## What is this Lightning Web Component thing?
```
```suggestion
## What is this Lightning Web Components thing?
```

malaporte commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+## What is this Lightning Web Component thing?
+
+Lightning Web Component includes two concepts in its name.
```
```suggestion
Lightning Web Components includes two concepts in its name.
```

malaporte commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+Lightning Web Component includes two concepts in its name.
+
+The **Lightning** part of the *Lightning Web Components*. Lightning is a Salesforce related term which defines their UI or front-end applicative framework (by opposition to "Classic" or "Visualforce"). 
```
```suggestion
The **Lightning** part is a Salesforce related term which defines their UI or front-end applicative framework (by opposition to "Classic" or "Visualforce"). 
```

malaporte commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+The **Lightning** part of the *Lightning Web Components*. Lightning is a Salesforce related term which defines their UI or front-end applicative framework (by opposition to "Classic" or "Visualforce"). 
+
+Second, **Web Components** are the result of three technologies that, when used in combination, allow you to define custom and reusable web elements. It leverages *Custom elements* which allows you to define custom web elements in JavaScript and some behaviour which in turn can be rendered in a web browser. These custom elements also use the *Shadow DOM* which is an encapsulation of the DOM tree  ...
```
Separate or isolate?

erocheleau replied:
the correct term is probably encapsulate :)

erocheleau replied:
But yea I like isolate better here.

malaporte commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+Before jumping on our experience with LWC, here are the members of our team that took part in the original exploration of this new framework.
+
+### Jeremie Robert
```
```suggestion
### Jérémie Robert
```

erocheleau replied:
Yea... I removed accent's from everyone's names. Do you think I should keep them?

André
Jérémie
Étienne

:)

malaporte replied:
As you wish I guess. Still triggers me :)

malaporte commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+### Jeremie Robert
+
+> At Coveo for over a year, I am constantly working at better integrating Coveo with the Salesforce platform. Using Salesforce technologies (Apex, Visualforce, Aura, and now LWC) and other web technologies (ES6, TypeScript, CSS, and SASS) I build solutions that can be personalized by our clients and that respect the highest standards in terms of quality.
```
no less 😛

erocheleau replied:
Nothing but the best

malaporte commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+> From a developer point of view, switching from Aura to LWC meant an access to better tools and following current web standards. We had put a lot of efforts to be able to have relatively good tests with Aura components, but with LWC it was part of the framework which in turn would mean a faster release cycle and higher product quality. To compare both, with Aura we could mainly do integration te ...
+> 
+> Also because I was already used to other web frameworks, I liked the simplicity of the boilerplate needed for LWC. Compared to Aura from a readability standpoint, LWC is better because of how the code is written. It makes the code clearer and clearer code means less mistakes.
```
```suggestion
> Also because I was already used to other web frameworks, I liked the reduced amount of boilerplate needed for LWC. Compared to Aura from a readability standpoint, LWC is better because of how the code is written. It makes the code clearer and clearer code means less mistakes.
```

malaporte commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+> 
+> Also because I was already used to other web frameworks, I liked the simplicity of the boilerplate needed for LWC. Compared to Aura from a readability standpoint, LWC is better because of how the code is written. It makes the code clearer and clearer code means less mistakes.
+
+\- Jeremie
```
```suggestion
\- Jérémie
```

malaporte commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+## What did you decide to create as your first experience with LWC?
+
+> Our Salesforce integration uses our own UI framework we have developed in house called the [Coveo Search-UI](https://github.com/coveo/search-ui). It is a collection of UI components all written in pure TypeScript that are used to build a search page. Our first component was a wrapper around this framework to integrate it within Salesforce and with LWC. This component is responsible of loading t ...
```
```suggestion
> Our Salesforce integration uses our own open source UI framework we have developed in house called the [Coveo Search-UI](https://github.com/coveo/search-ui). It is a collection of UI components all written in pure TypeScript that are used to build a search page. Our first component was a wrapper around this framework to integrate it within Salesforce and with LWC. This component is responsible of loading the required Static Resources and then initializing our framework.
```

malaporte commented on _posts/2019-04-18-lightning-web-components.md:
````diff
@@ -0,0 +1,205 @@
+```
+> You clearly see what is exposed publicly outside of your component. It makes it easy to use composition of components, it is similar to React, but uses the shadow DOM to process rendering and to encapsulate behavior and data. And the `@wire` decorator makes it very simple to understand clearly that you are calling Apex declaratively and letting Salesforce take care of caching the information.
+
+\- Jeremie
````
should I bother?

malaporte commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+\- Louis
+
+> Quote from Andre about previous pain points around not using sfdx/Aura components.
```
is that a TODO?

amoreauCoveo replied:
Upvote

erocheleau replied:
Commit mixed up, I had removed it in the other branch I guess... thanks.

amoreauCoveo commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+\- Andre
+
+> It feels very natural to develop with LWC. It seems like the design decisions make sense and the integration with Salesforce's backend feels seamless. I am sure it will be easier to ramp up a new member of the team with LWC. One of the best lesson learned over the years is to *keep it simple* and LWC lets us do that. We can have clearly defined building blocks (components) that are designed to  ...
```
One of the best lessons (missing s)

Add comma between Javascript and "and CSS"

You reversed the brackets for the Search UI link ( should be `[name](link)` )

Remove the "And" at the beginning of the last sentence

erocheleau replied:
Just for you :) Otherwise nobody would understand then sentence :)

amoreauCoveo commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+## What will the future look like for the Coveo for Salesforce integration team?
+
+> The Coveo for Salesforce integration package grew a lot over the past years. We've obviously added a lot of new features, but with the transition from Visualforce to Aura, and now the transition to LWC, our package is full of relics. We look forward to committing more resources to LWC and phasing out more and more Visualforce bit by bit. With LWC being closer to web standards than either Visual ...
```
than either Visualforce or Aura (not and)

amoreauCoveo commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+## Would you like to know more?
+
+### Read more on [Web Components](https://developer.mozilla.org/en-US/docs/Web/Web_Components) (MDN)
```
I don't think those need to be headings

amoreauCoveo commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+### Read more on [Web Components](https://developer.mozilla.org/en-US/docs/Web/Web_Components) (MDN)
+
+### Read more on [Lightning Web Components](https://developer.salesforce.com/docs/component-library/documentation/lwc/lwc.get_started_introduction) (developer.salesforce)
```
developer.salesforce.com looks less weird

erocheleau replied:
Yea but it renders as an url if I add a .com...

Changed it to (Salesforce developers) which is their "title"

amoreauCoveo commented:
To visualize the blog post: https://erocheleau.github.io/source.coveo.com/2019/04/18/lightning-web-components/

erocheleau commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+The **Lightning** part of the *Lightning Web Components*. Lightning is a Salesforce related term which defines their UI or front-end applicative framework (by opposition to "Classic" or "Visualforce"). 
+
+Second, **Web Components** are the result of three technologies that, when used in combination, allow you to define custom and reusable web elements. It leverages *Custom elements* which allows you to define custom web elements in JavaScript and some behaviour which in turn can be rendered in a web browser. These custom elements also use the *Shadow DOM* which is an encapsulation of the DOM tree  ...
```
I was trying not to have a formal 1. 2. 3. basically

IMO there are much better sources of info on what a Web component is and I am trying to keep it at bare minimum here, see the first link at the bottom of the post.

erocheleau commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+### Jeremie Robert
+
+> At Coveo for over a year, I am constantly working at better integrating Coveo with the Salesforce platform. Using Salesforce technologies (Apex, Visualforce, Aura, and now LWC) and other web technologies (ES6, TypeScript, CSS, and SASS) I build solutions that can be personalized by our clients and that respect the highest standards in terms of quality.
```
Maybe, I thought about the pictures, didn't want to make it too long.

erocheleau replied:
I really wanted the quotes since it was the own words from every team member.
Any other idea to keep quotes but no bold? indeed it makes it look a lot better in regards to style if I remove the bold.

erocheleau commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+### Etienne Rocheleau
+
+> At Coveo for 4 years, my job consists of the exploration of all new technology from Salesforce. I am always looking for new ways to integrate Coveo with Salesforce to create a better experience for users. I produce a lot of demos, POCs, and prototypes as part of my exploration and I really love when people find new and interesting ways to use Coveo to solve an issue they have. I am very familia ...
```
I'm trying okay... ;)

erocheleau commented on _posts/2019-04-18-lightning-web-components.md:
````diff
@@ -0,0 +1,205 @@
+
+\- Jeremie
+
+```js
````
I tried something with it, check updated PR please.

erocheleau commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+> Above is an example of our first LWC component. It loads the CSS and JS files we need on the `connectedCallback` which is a lifecycle method that is called on component initialization by Salesforce. And then on the `renderedCallback` it initializes our framework, this method is also called by Salesforce when a component is rendered (our framework is designed to only initialize itself once). 
+>
+> Unfortunately, we cannot rely on some of the DOM improvements provided by the shadowDOM since we handle our own rendering loop of our components. This is done in the template file by activating `lwc:dom="manual"` on our element container. But since we are already pure JavaScript and following web design principles, it is still really fast.
```
Changed to 
> we handle ourselves the rendering loop of our components

erocheleau commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+\- Louis
+
+> From the original exploration of LWC, I found that it made a lot of sense to use the declarative approach from the LWC framework. It is very clear, simple and elegant the way you describe a component's *API*. 
```
Does it matter if it is, clear + (simple and elegant) or clear + simple + elegant? ;)

erocheleau commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,205 @@
+
+![Coveo SFINT folder structure]({{ site.baseurl }}/images/2019-04-15-LWC/project_structure.png)
+
+> With the new advantages provided by `sfdx` but also the improved intellisense from the LWC framework, we now have reason enough to look at making the switch to the Salesforce folder structure instead. This new structure simplifies a lot our build process because we can remove a lot of the manual files handling with Gulp. I still feel however that more freedom with how the files are organized co ...
```
No to your 2nd point, Gulp is used to handle files manually, but not causing it.

changed it to :
> This new structure simplifies a lot our build process because we can remove a lot of the manual files handling we were doing with Gulp.

erocheleau commented:
In the end, I changed the CSS to remove the default bold on quotes. It was not looking really good.
I took a quick look in the other blog posts, and it looks fine too when people are using it.

You can still do a bolded quote if you want by doing `> ** text **`

I have modified the Facet blog post from @acverrette because these looked better as bolded.

malaporte commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,203 @@
+
+## What does this Lightning Web Components name mean?
+
+Lightning Web Components include two concepts in its name.
```
```suggestion
Lightning Web Components include two concepts in its name:
```

malaporte commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,203 @@
+
+So if you put the two together, you get Web Components that can interact with Salesforce, either to access data or to be part of their application rendering loop. Which means Salesforce has made available this new bleeding edge web technology (Web Components) and allowed it to communicate very easily with Salesforce (Lightning).
+
+## Meet the team
```
What no pictures? :)

malaporte commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,203 @@
+
+In 2018, Salesforce introduced a new programming model for building Lightning components. As you may already know, Lightning components are the building blocks used to create applications within Salesforce. This new framework was designed with respect to the new web standards in order to allow web developers to keep using the knowledge they have acquired over the years outside of Salesforce. This ...
+
+<!-- more -->
```
I think you should do a bit more to explain that this post is written as an interview of the developers. Right now you hint at this only in the last sentence of a big paragraph, and even then it's not so obvious. I know that I was a bit surprised by the format afterwards when reading.

malaporte commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,203 @@
+
+> At Coveo for 4 years, and my job consists of the exploration of all new technology from Salesforce. I am always looking for new ways to integrate Coveo with Salesforce to create a better experience for users. I produce a lot of demos, POCs, and prototypes as part of my exploration and I really love when people find new and interesting ways to use Coveo to solve an issue they have. I am very fam ...
+
+## Why switch to LWC when you are so involved with Aura?
```
The CSS doesn't make a big difference between the heading levels - it makes the structure a bit harder to see IMO.

erocheleau replied:
Yea resolved that with a css proposed change. All quotes were in **bold** by default, removed that styling. It looks better now.

pmdartus commented on _posts/2019-04-18-lightning-web-components.md:
```diff
@@ -0,0 +1,203 @@
+> Some gotchas to take into account that had me searching for a little while:
+>
+> 1. As of writing this post, LWC does not allow for custom decorators, only the ones provided by Salesforce and the LWC framework. JavaScript decorators are still only a proposition to ECMAScript, but seeing the decorators offered by Salesforce makes me want to be able to create my own. For example, I would like to create a `@log({level: 'info'})` decorator to mark a method to log to the console ...
+> 2. Be wary of the difference between `this.querySelector`, `this.template.querySelector`, and `element.shadowRoot.querySelector`. The first one is used to search the DOM in a *slot*, the second one is used to search within the body of your LWC, and the third one is used within unit tests.
```
`element.shadowRoot.querySelector` is equivalent to `this.template.querySelector` and `element.querySelector` is equivalent to `this.querySelector`. The first queries the shadow DOM and the later queries the light DOM. It's probably something we should enhance in our documentation.

Maybe this section should be reformulated not to confuse readers.

erocheleau replied:
Okay this was based on our experience and something we saw in documentation around unit tests with Jest.
You are right, after looking at examples, we were only using `element.shadowRoom.querySelector` because we are creating ourselves the element with 
`const element = createElement();`
Thank you, I will probably remove that part since `element.` is confusing indeed.
I still like the difference between the `this.querySelector` and `this.template.querySelector` though, what do you think?
It comes from here:
https://developer.salesforce.com/docs/component-library/documentation/lwc/create_components_slots

erocheleau commented:
Closing this as this is not really apropos anymore 👎
